### ISSUE TITLE
Initialization-order problem

### ISSUE THREAD

**Issue by Remi-Coulom** [kripken/emscripten#4464]
I am using WebIDL according to the [documention](https://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/WebIDL-Binder.html), and incorporate the glue code into the emscripten output as recommended, with the `--post-js glue.js` option.

This works very well with optimized builds, but not with `-O0`. With no optimization, the main function is called before the post-js code is executed, so the glue is not available.

Is there a way to make things work whatever the optimization level? I could manually fix the problem by moving the call to `run()` at the end of the emscripten output (after the post-js code).

I feel run() should always be after the post-js code, or there should be a way to insert js code between the end of the module definition and the call to run(). I tried hard to find an option, but could not find one. Is there a solution to make things work properly whatever the optimization level?

**Comment by kripken** [comment]
The main difference is that in optimized builds we have a side file with the mem init. We must wait for an async load for it to arrive.

You can just disable that (`--memory-init-file 0`) and startup will by sync. The cost is code size for non-binary mem init.

Alternatively, the more correct approach is to not put code in a post-js that runs unconditionally. Instead, you can put it in a function. And have the function called after startup is complete (e.g. by calling it from `main()`, or by putting it on `Module.postRun`).

**Comment by Remi-Coulom** [comment]
Thank you very much for your feedback.

I'd like to use the "more correct approach", but is there a simple way to generate WebIDL output that works with the "more correct approach"? I found some ways to include the js code in [that stackoverflow question](http://stackoverflow.com/questions/950087/how-to-include-a-javascript-file-in-another-javascript-file). It is a bit complicated for my taste.

The solution I am using now is to always compile with optimization. But I am wondering: do I have the guarantee that the onRuntimeInitialized event will take place after the post-js code is run? Is the async load always delayed? If it is always delayed, then it is OK. Otherwise, I'll have to find another solution.

**Comment by kripken** [comment]
Yes, you have a guarantee that `main()` (and `postRun`s, and `onRuntimeInitialized`) can only be called when everything is ready, which includes async operations like mem init file (if there is one), preloaded files (if there are any).

See also [this faq entry](http://kripken.github.io/emscripten-site/docs/getting_started/FAQ.html#how-can-i-tell-when-the-page-is-fully-loaded-and-it-is-safe-to-call-compiled-functions).

**Comment by Remi-Coulom** [comment]
Unfortunately, this is not the case with -O0: `main` and `onRunTimeInitialized` are called _before_ the --post-js code is run. It is a bit annoying that the guarantee depends on optimization level.

**Comment by kripken** [comment]
Yes, perhaps I wasn't clear. Let me try again. The code looks like this

````
pre-js
main
post-js
````

If main is ready to go - which is the case if there is nothing async - then it runs. And the post-js code is physically after, so it runs after.

if main is **not** ready to go - if there is a mem init file we must wait for asyncly - then main will not run. post-js will then be run since it is there **but this is invalid** (and a build with assertions will show an error, as mentioned in that faq entry - if we aren't ready to run compiled code, post-js code that calls compiled code is bad).

The proper solution is to run code when it is ready, calling it from main or postRun etc. for example, put your code in a pre-js **function**:

````
function myCode() {
}
main, and main calls myCode()
````

then main can be called either immediately or later, and when it is ready, it will call your code, and everything should work ok.

**Comment by Remi-Coulom** [comment]
OK, thanks, I understand.

**Comment by juj** [comment]
Hi @Remi-Coulom, great to see you here, I'm a big fan!

This reads like it can be closed as resolved, but if you have any other issues or questions, don't hesitate to open a new item.

**Comment by kripken** [comment]
Thanks, we should fix those docs. But reading them now I'm not sure where, `--post-js` is mentioned for including the glue code, not user code. Is that what was confusing, or was it somewhere else?

**Comment by Remi-Coulom** [comment]
On [that page](https://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/WebIDL-Binder.html):
> Add --post-js glue.js in your final emcc command. The post-js option adds the glue code at the end of the compiled output.
This will be executed after main with -O0

**Comment by kripken** [comment]
I see, thanks. I'll clarify that text better.

**Comment by Remi-Coulom** [comment]
My feeling is that it might be a better idea to simply insert all the post-js code right before the call to run(), instead of after it. This way the post-js code will have the same execution order with respect to main, whatever the optimization level. But I am not extremely familiar with emscripten, so my suggestion might have bad consequences in some other situation I can't think of.

**Comment by kripken** [comment]
Yes, it's tricky, since one usage of `--pre-js` and `--post-js` is to wrap the entire codebase in a function. That depends on them literally being placed before and after the main emitted code.

I updated the docs which I hope will help here. But I also like your suggestion of `--post-js-but-before-run`, that would basically wrap the provided code in a `postRun`. Perhaps it could be literally `--post-run`?
